storage/registers.go

```diff
@@ -0,0 +1,33 @@
+package storage
+
+import (
+	"io"
+
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// Registers defines methods for the register index.
+type Registers interface {
+	RegisterReader
+	RegisterWriter
+	io.Closer
+}
+
+// RegisterReader defines read-only operations on the register index.
+type RegisterReader interface {
+	// Get register by the register ID at a given block height.
+	//
+	// If the register at the given height was not indexed, returns the highest
+	// height the register was indexed at.
+	// An error is returned if the register was not indexed at all. Expected errors:
+	// - storage.ErrNotFound if the register was not found in the db
+	Get(ID flow.RegisterID, height uint64) (flow.RegisterEntry, error)
+}
+
+// RegisterWriter defines write-only operations on the register index.
+type RegisterWriter interface {
+	// Store batch of register entries at the provided block height.
+	//
+	// If the registers already exists it overwrites them to make this action idempotent.
```

Comment:
I don't think the storage interface needs to make this promise, just the indexer. It may actually be valuable to prevent overwriting existing entries after they are committed, but I'm not 100% sure if that's necessary here.

re: my previous comment, there may not be any expected errors at this point. we can add any that come up later if necessary.

Reply:
I wonder how would the indexer make that possible if the storage layer errors on overwriting? My thinking for defining that as part of the interface is because it's not just an implementation detail but the behavior of the API. I can remove the comment, the only thing I would like to understand is how would then the indexer guarantee this logic in case we can not make a single batch update?

Reply:
There is additional error to handle: the caller must call it with the height in strict increasing order. Skipping a height should return a fatal error as it indicates a bug, and the caller should just crash.

Reply:
Agree on the sentinel error.

About the increasing order, I'm wondering if that's the job of a storage layer. It feels to me (and I trust your judgment saying otherwise, I just wonder about it), that it couples the storage logic with indexing domain logic. The indexer module it feels to me should guarantee updates increasing height by 1, more specifically I was thinking of using the jobqueue to provide just that functionality. @zhangchiqing